huggingface/accelerate
# Allow for samplers to be seedable and reproducable

# What does this PR do?

This PR ensures that `DataLoaders` made through `Accelerate` are reproducible on distributed systems by allowing `set_epoch` to work on `RandomSamplers` (what gets selected if `shuffle=True` on a `DataLoader`).

Fixes # (issue)

fixes https://github.com/huggingface/transformers/issues/26541
fixes https://github.com/huggingface/accelerate/issues/2016

To test this works, I ran the following script:

```python
from accelerate import Accelerator
from accelerate.utils import set_seed

set_seed(42)
accelerator = Accelerator()
dataset = range(24)
dataloader = torch.utils.data.DataLoader(dataset, batch_size=8, shuffle=True)
dataloader = accelerator.prepare(dataloader)

dataloader.set_epoch(0)
original_outputs = []

for x in dataloader:
    original_outputs += x

original_outputs = torch.stack(original_outputs)

dataloader.set_epoch(0)
new_outputs = []

for x in dataloader:
    new_outputs += x
    
new_outputs = torch.stack(new_outputs)

print(f'Original: {original_outputs}\n')
print(f'New: {new_outputs}\n')
```

When comparing the outputs, these two line up perfectly, ensuring we can in-fact recreate what was originally setup.

For testing we will be relying on the transformers integration CI (see the test runners on main that include the failing test). Linked with https://github.com/huggingface/transformers/pull/26850

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
      to it if that's the case.
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@BenjaminBossan @LysandreJik
